addon/mixins/copyable.js

```diff
@@ -1,136 +1,181 @@
 import Ember from 'ember';
 import DS from 'ember-data';
 
-export default Ember.Mixin.create({
+const {
+  A,
+  get,
+  set,
+  Mixin,
+  typeOf: emberTypeOf,
+  RSVP: { Promise, allSettled }
+} = Ember;
+
+const {
+  PromiseObject,
+  PromiseManyArray
+} = DS;
+
+let failSafeAll = (array) => {
+  return allSettled(array).then(response => {
+    return A(response || []).mapBy('value').filter(e => e);
+  });
+};
+
+export default Mixin.create({
   copyable: true,
-  copy: function(options, copied) {
+
+  copy(options) {
     options = options || {};
-    copied = copied || {};
+    return this._copyAsync(options, {}).
+                then((copy) => {
+                  return this._applyOptions(options, copy);
+                });
+  },
+
+  shallowCopy(options) {
+    options = options || {};
+    return this._copyAsync(options, {}, true).
+                then((copy) => {
+                  return this._applyOptions(options, copy);
+                });
+  },
+
+  _applyOptions(options, copy) {
+    if(!copy) {
+      return;
+    }
+
+    for(let optionName in options) {
+      let option = options[optionName];
+      if(emberTypeOf(option) === 'object') {
+        this._applyOptions(option, get(copy, optionName));
+      } else {
+        set(copy, optionName, option);
```

Comment:
Could you fork my forked repo and provide a test? I'll have a look into it then.